[TS-1215] Correct work with internal state (#7)
th2-net/th2-read-pcap-fix: README.md

> # read pcap fix (version)

Reviewer: ```suggestion
# read pcap fix (0.0.2)
```

## Overview

> This component is designed to read FIX packets from pcaps and publish them as raw messages.

Reviewer: ```suggestion
This component is designed to read FIX messages from pcaps and publish them as raw messages.
```

In case of shutdown, the process can be resumed from the point where it was interrupted.

> States, which are stored after a certain interval, are used for that.

Reviewer: Could you please write that state is stored on the disk and PV is required for this component.

## Configuration parameters

All values after `:` are default values.

> **use_mstore: _false_** — the flag that is responsible for saving messages to 

Reviewer: ```suggestion
**use_mstore: _true_** — the flag that is responsible for saving messages to 
```

[mstore](https://github.com/th2-net/th2-mstore) instead of a database

**checkpoint_interval: _10000_** — the number of packets after which the state will be saved

> **dead_connections_scan_interval: _100000_** — this parameter specifies the time after which the scanning of dead 

Reviewer: What time unit is used?

connections is run

> **connection_death_interval: _10800000_** — this parameter specifies the time after which the connection 

Reviewer: What time unit is used?
Also, please clarify what time is used by component: realtime or pcap time.

is considered dead

**read_state: _true_** — this parameter specifies if the component reads its state when (re)staring in order to 
resume the processing

**write_state: _true_** — this parameter specifies if the component write its state which is used 
to resume the processing

**use_timestamp_from_pcap: _true_** — this parameter specifies if timestamps of the raw messages will be set 
as they are in PCAPs (true) or as the current time on the server (false)

> **sleep_interval: _60_** — number of seconds(?) that the component will wait after finishing processing of all 

Reviewer: `seconds(?)` please clarify

available packets in PCAPs before
checking if there's more to be processed

> **number_of_packets_to_successful_restore: _10_** — the number of packets needed to successfully restore a gap 

Reviewer: How it works?

Reviewer 2: i haven't quite grasped the logic behind it, here's the code (`FIXReader.restoreStream(boolean isClientStream,Pcap associatedPacket)`):
```
if (successfulPackets >= configuration.getNumberOfPacketsToSuccessfulRestoreFIX()) {
            resetState(false);
            this.restoreMode = false;
            return packetsToRestore;
        }
```

I don't get how a certain amount of packets is necessary for restoring a stream.

Reviewer: I have tried to review code, but I have not understood yet. I will try late again.
@OptimumOpium , maybe you have idea about this option?

in a stream message stream

**use_event_publish: _true** — this parameter specifies if the component will publish any events

> **disable_cradle_saving: _false_** — this parameter specifies if the component will store the messages to 

Reviewer: Is it related to the use_mstore option ?

Reviewer 2: Seems like this parameter does duplicate the `use_mstore` logic, but those are used in different places. Not sure if it's safe to replace it with `use_mstore`.

Reviewer: Active saver (MstoreFixSaver or CradleFixSaver) depends on use_mstore settings. disable_cradle_saving allow to control CradleFixSaver behavior. Looks like dev has added it for internal needs.
Let's describe it in readme

[cradle](https://github.com/th2-net/cradleapi)

> **disable_connectivity_saving: _false_** — this parameter specifies if the connection event will be stored

Reviewer: Maybe rename to connectivity messages

resume the processing

**write_state: _true_** — this parameter specifies if the component write its state which is used 
to resume the processing

**use_timestamp_from_pcap: _true_** — this parameter specifies if timestamps of the raw messages will be set 
as they are in PCAPs (true) or as the current time on the server (false)

**sleep_interval: _60_** — number of seconds(?) that the component will wait after finishing processing of all 
available packets in PCAPs before
checking if there's more to be processed

**number_of_packets_to_successful_restore: _10_** — the number of packets needed to successfully restore a gap 
in a stream message stream

**use_event_publish: _true** — this parameter specifies if the component will publish any events

**disable_cradle_saving: _false_** — this parameter specifies if the component will store the messages to 
[cradle](https://github.com/th2-net/cradleapi)

**disable_connectivity_saving: _false_** — this parameter specifies if the connection event will be stored

**event_batch_size: _1048576_** — this parameter specifies the size of an event batch (1MB by default)

> **event_batcher_core_pool_size: _2_** —  this parameter specifies the number of events necessary for sending another 

Reviewer: ```suggestion
**event_batcher_core_pool_size: _100_** —  this parameter specifies the number of events necessary for sending another 
```

Reviewer 3: Why do we call it `pool_size`? In my understanding start sounds like something related to the thread inside the application

event batch

**event_batcher_max_flush_time: _1000_** — the number of milliseconds for scheduling the sending of event batches

**use_offset_from_cradle: _true_** — this parameter is used for synchronization of PCAP packet sequences with
packet sequences from Cradle

> **individual_read_configurations** — this parameter will be covered in the next section

Reviewer: this parameter will be covered in the next section
What is it mean ?

> **message_batcher_core_pool_size: _2_** — this parameter specifies the number of events necessary for sending another

Reviewer: ```suggestion
**message_batcher_core_pool_size: _100_** — this parameter specifies the number of messages necessary for sending another
```

message batch

**message_batcher_max_flush_time: _1000_** — the number of milliseconds for scheduling the sending of message batches

> **buffered_reader_chunk_size: _8192_** — the buffer size for BufferedInputStream

Reviewer: Could you please explain what does it mean for a user

> **check_message_batch_sequence_growth: _false_** — this parameter specifies if the component checks whether 

Reviewer: ```suggestion
**check_message_batch_sequence_growth: _true_** — this parameter specifies if the component checks whether 
```

sequence numbers are incremental in message batches

> **check_message_batch_timestamp_growth: _false_** — this parameter specifies if the component checks whether

Reviewer: ```suggestion
**check_message_batch_timestamp_growth: _true_** — this parameter specifies if the component checks whether
```

timestamps are incremental in message batches 

**tcpdump_snapshot_length: _262144_** — this parameter specifies the maximum packet length

**sort_th2_messages: _false_** — this parameter specifies if messages should be sorted by timestamps

> **usable_fraction_of_batch_size: _0.9_** — the fraction of batch size to be used; can be from 0 to 1

Reviewer: Could you please write the explanation why this parameter is added

Reviewer 2: I don't remember the details, but there were problems occasionally with processing the batches by the Processor if this parameter was set to 1.0.

Reviewer: The main problem is a deterrent logic to calculate message size in read and mstore.

**message_sorter_window_size: _15000_** — the time window size in which a sorting is performed

**message_sorter_connection_end_timeout: _30000_** — this parameter specifies the amount of time (in milliseconds)
that is supposed to pass between two adjacent packets so the connection would be considered stale and then deleted

**message_sorter_clear_interval: _2000_** — the amount of messages used by MessageSorter to start searching for
stale sessions (i.e. after sorting the next 2000 messages the search begins)

**possible_time_window_between_files: _1000_** — this parameter specifies the **maximum** amount of time (in milliseconds)
that is supposed to pass between two adjacent PCAPs 
(the last packet timestamp for 1st PCAP and the first packet timestamp from the 2nd PCAP are compared).

> **check_message_size_exceeds_batch_size: _false_** — this parameter specifies if a message size exceeding a batch size 

Reviewer: ```suggestion
**check_message_size_exceeds_batch_size: _true_** — this parameter specifies if a message size exceeding a batch size 
```
